### ISSUE TITLE
Rabbit MQ connection lost leads to application shutown

### ISSUE THREAD

**Issue by adunne-resilient** [guidesmiths/rascal#93]
When the connection to RabbitMQ is lost, my application is shutting down with the error: "CONNECTION_FORCED - broker forced connection closure with reason 'shutdown'". However, it seems that if I first publish a message to a publication, the application will stay alive and try to reconnect.

My configuration is:

```
const config: rascal.BrokerConfig = {
  vhosts: {
    shovel: {
      connection: {
        url: 'amqp://localhost:5672/shovel?hearbeat=5'
      },
      queues: ['testQueue'],
      publications: {
        demoPub: {
          queue: 'testQueue'
        }
      },
      subscriptions: {
        demoSub: {
          queue: 'testQueue',
          prefetch: 3
        }
      }
    }
  }
};
```

Example code:

```
const broker = await rascal.BrokerAsPromised.create(rascal.withDefaultConfig(config));
    broker.on('error', (err) => {
      console.log('error on broker!', err);
    });

    // Publish a message
    const publication = await broker.publish('demoPub', 'Hello World!'); // required to keep connection alive it seems
    publication.on('error', () => console.error('error on publication!'));

    // Consume a message
    const subscription = await broker.subscribe('demoSub', { prefetch: 10, retry: false });
    subscription
      .on('message', (message: rascal.Message, content: Buffer, ackOrNack: rascal.AckOrNackFn) => {
        console.log('Received...', content.toString());
        ackOrNack();
      })
      .on('error', (err) => console.error('error on subscription!', err));
`

**Comment by cressie176** [comment]
Hi @adunne-resilient,

I suspect the problem is that there's nothing to keep the node process alive once the connection has closed. As a quick test can you try adding the following code to your application.

```js
setInterval(() => {}, 60000);
```

Some alternative methods are suggested [here](https://stackoverflow.com/questions/23622051/how-to-forcibly-keep-a-node-js-process-from-terminating)

**Comment by adunne-resilient** [comment]
Yes, that keeps the node process alive, but when the rabbitMQ server comes back up the connection is not reestablished, and messages are not processed until I restart the application. This is the case even if  I have previously published to a publication, as mentioned above.

Is there something missing from my config above relating to reconnecting?

**Comment by cressie176** [comment]
```
const subscription = await broker.subscribe('demoSub', { prefetch: 10, retry: false });
```

You're overriding the configuration with `{ prefetch: 10, retry: false }` for this subscription. This will prevent retrying the subscription following an error.

**Comment by adunne-resilient** [comment]
Yes, this works if I remove the "retry: false". Am I missing a similar option on the broker connection? Or do I need to handle 'shutdown' events in some way?

It works if I include "setInterval(() => null, 60000);", I'm just wondering if this is necessary because my other config is incorrect.

Thank you for the help.

**Comment by cressie176** [comment]
> Am I missing a similar option on the broker connection? Or do I need to handle 'shutdown' events in some way?

There's nothing in the rascal config for keeping your process busy, so you will need to keep the process active in some way, although I'm a little surprised this hasn't come up before. I recently noticed that I wasn't always calling `unref()` after `setTimeout()`, so it's possible fixing this has introduced the problem. It's not something I would notice normally, as all the apps I write include an http server for reporting application health, and this would keep the process alive.

I think it's reasonable for rascal to do something like
```
setInterval(noop, A_VERY_LARGE_NUMBER)
```
And unref this when the broker is explicitly shutdown. I try to make this change over the weekend.

**Comment by adunne-resilient** [comment]
Ah yes, I see. I am just running this as a single function in a typescript file. I've seen it used previously as part of an http server, so, as you say, it wouldn't become an issue there.

Thanks again for the very useful information.

**Comment by cressie176** [comment]
Published as rascal@9.2.0
